pkg/federatedtypes/service.go

```diff
@@ -107,6 +107,28 @@ func (a *FederatedServiceAdapter) ObjectForCluster(template, override pkgruntime
 	return service
 }
 
+func (a *FederatedServiceAdapter)ObjectForUpdateOp(desiredObj, clusterObj pkgruntime.Object) pkgruntime.Object {
+	desiredService := desiredObj.(*corev1.Service)
+	clusterService := clusterObj.(*corev1.Service)
+
+	// ClusterIP and NodePort are allocated to Service by cluster, so retain the same if any while updating
+	desiredService.Spec.ClusterIP = clusterService.Spec.ClusterIP
+	for _, cPort := range clusterService.Spec.Ports {
+		for i, fPort := range clusterService.Spec.Ports {
+			if fPort.Name == cPort.Name && fPort.Protocol == cPort.Protocol && fPort.Port == cPort.Port {
+				desiredService.Spec.Ports[i].NodePort = cPort.NodePort
```

Comment:
I don't think it can be assumed that the port layout of the desired and cluster services will match.  In the event of divergence between the 2, this line seems to introduce the possibility of an index error.

Reply:
oops, there is a bug in line no.117. it should have been desiredService instead of clusterService. Will fix it. Basically i am iterating through all ports in clusterService to find a matching port in desiredService. if there is a match then only the NodePort in desiredService is retained to the same as that of clusterService port. Let me know if that's ok, after the fix.